### PR TITLE
acquisition: receipt detail view & refactoring

### PR DESCRIPTION
**Pull request by zannkukai**
Adds a component to display all details about a receipt.
Refactors acquisition modules services : removes generic services to
keep only API services ; removes all unnecessary classes in favor of
corresponding interfaces (lighter for memory usage).

Co-authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: projects/admin/src/app/acquisition/api/acq-account-api.service.ts [hunk 1]
```diff
@@ -42,14 +68,31 @@ export class AcqAccountApiService {
...
+  /**
+   * Get account from ES (metadata from ES much more complete than DB stored record)
+   * @param accountPid - the account pid.
+   * @returns the corresponding account
+   */
+  getAccount(accountPid: string): Observable<IAcqAccount> {
+    return this._recordService
+      .getRecords(this.resourceName, `pid:${accountPid}`, 1, 1)
+      .pipe(
+        map((result: Record) => this._recordService.totalHits(result.hits.total) === 0 ? [] : result.hits.hits),
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-account-api.service.ts:82]: just return `result.hits.hits`

File: projects/admin/src/app/acquisition/api/acq-account-api.service.ts [hunk 2]
```diff
@@ -61,7 +104,11 @@ export class AcqAccountApiService {
     const query = defaultQueryParams.join(' AND ');
     options = { ...{sort: 'name'}, ...options };  // add some default params
     return this._recordService
-      .getRecords(this.resourceName, query, 1, RecordService.MAX_REST_RESULTS_SIZE, undefined, undefined, undefined, options.sort);
+      .getRecords(this.resourceName, query, 1, RecordService.MAX_REST_RESULTS_SIZE, undefined, undefined, undefined, options.sort)
+      .pipe(
+        map((result: Record) => this._recordService.totalHits(result.hits.total) === 0 ? [] : result.hits.hits),
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-account-api.service.ts:109]: just return `result.hits.hits`

File: projects/admin/src/app/acquisition/api/acq-budget-api.service.ts [hunk 3]
```diff
@@ -61,8 +66,8 @@ export class AcqBudgetService {
       .getRecords('acq_accounts', `budget.pid:${budgetPid} AND depth:0`, 1, RecordService.MAX_REST_RESULTS_SIZE)
       .pipe(
         map((result: Record) => this._recordService.totalHits(result.hits.total) === 0 ? [] : result.hits.hits),
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-budget-api.service.ts:68]: just return `result.hits.hits`

File: projects/admin/src/app/acquisition/api/acq-order-api.service.ts [hunk 4]
```diff
@@ -46,41 +85,67 @@ export class AcqOrderApiService {
...
+  getOrderLines(orderPid: string, extraQuery?: string): Observable<IAcqOrderLine[]> {
     let query = `acq_order.pid:${orderPid}`;
     if (extraQuery) {
       query += ` ${extraQuery}`;
     }
     return this._recordService
-      .getRecords('acq_order_lines', query, 1, RecordService.MAX_REST_RESULTS_SIZE, undefined, undefined, undefined, 'priority');
+      .getRecords('acq_order_lines', query, 1, RecordService.MAX_REST_RESULTS_SIZE, undefined, undefined, undefined, 'priority')
+      .pipe(
+        map((result: Record) => this._recordService.totalHits(result.hits.total) === 0 ? [] : result.hits.hits),
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-order-api.service.ts:130]: just return `result.hits.hits`

File: projects/admin/src/app/acquisition/api/acq-receipt-api.service.ts [hunk 5]
```diff
@@ -17,37 +17,66 @@
...
+import { AcqResponseReceiptLineStatus, ICreateLineMessage, IResponseReceiptLine } from '../components/receipt/receipt-form/order-receipt';
 
 @Injectable({
   providedIn: 'root'
 })
 export class AcqReceiptApiService {
 
+  // SERVICES ATTRIBUTES ======================================================
+  /** The resource name for an acquisition receipt */
+  resourceName = 'acq_receipts';
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-receipt-api.service.ts:36]: `readonly`

File: projects/admin/src/app/acquisition/api/acq-receipt-api.service.ts [hunk 6]
```diff
@@ -73,40 +102,80 @@ export class AcqReceiptApiService {
...
     headers?: object,
     sort?: string
-  }): Observable<any|null> {
+  }): Observable<IAcqReceipt[]> {
     options = {...this._defaultRecordOptions, ...options};  // add some default params
     return this._recordService
-      .getRecords(this.resourceName, query, 1, RecordService.MAX_REST_RESULTS_SIZE, undefined, undefined, options.headers, options.sort);
+      .getRecords(this.resourceName, query, 1, RecordService.MAX_REST_RESULTS_SIZE, undefined, undefined, options.headers, options.sort)
+      .pipe(
+        map((result: Record) => this._recordService.totalHits(result.hits.total) === 0 ? [] : result.hits.hits),
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-receipt-api.service.ts:110]: just return `result.hits.hits`

File: projects/admin/src/app/acquisition/api/acq-receipt-api.service.ts [hunk 7]
```diff
@@ -73,40 +102,80 @@ export class AcqReceiptApiService {
...
+   * @param receiptPid - the parent receipt pid
+   * @returns an observable of AcqReceiptLines sorted on `receipt_date`
+   */
+  getReceiptLines(receiptPid: string): Observable<IAcqReceiptLine[]> {
+    const query = `acq_receipt.pid:${receiptPid}`;
+    return this._recordService
+      .getRecords('acq_receipt_lines', query, 1, RecordService.MAX_REST_RESULTS_SIZE,
+                  undefined, undefined, BaseApi.reroJsonheaders, 'receipt_date')
+      .pipe(
+        map((result: Record) => this._recordService.totalHits(result.hits.total) === 0 ? [] : result.hits.hits),
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/api/acq-receipt-api.service.ts:135]: just return `result.hits.hits`

File: projects/admin/src/app/acquisition/classes/common.ts [hunk 8]
```diff
@@ -19,44 +19,33 @@
...
 
-/** Interface to describe an order note */
-export interface AcqNote {
-  type: AcqNoteType;
-  content: string;
+export interface IObjectReference {
+  $ref?: string;
+  pid?: string;
+  type?: string;
 }
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/classes/common.ts:51]: Candidate for the shared ?

File: projects/admin/src/app/acquisition/classes/order.ts [hunk 9]
```diff
@@ -31,7 +30,7 @@ export interface AcqAddressRecipient {
 
 /** Enumeration about order status */
 export enum AcqOrderStatus {
-  CANCELED = 'canceled',
+  CANCELLED = 'cancelld',
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/classes/order.ts:33]: Missing `e`

File: projects/admin/src/app/acquisition/components/order/order-brief-view/order-brief-view.component.ts [hunk 10]
```diff
@@ -46,23 +47,58 @@ export class OrderBriefViewComponent implements ResultItem, OnInit {
...
+
+  }
+
+
+  // COMPONENTS FUNCTIONS =====================================================
+  /**
+   * Is the encumbrance data should be visible or not.
+   * @returns True if the encumbrance accounting data could be visible.
+   */
+  displayEncumbranceAccountingData(): boolean {
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/order/order-brief-view/order-brief-view.component.ts:90]: `get displayEncumbranceAccountingData(): boolean {` ?

File: projects/admin/src/app/acquisition/components/order/order-brief-view/order-brief-view.component.ts [hunk 11]
```diff
@@ -46,23 +47,58 @@ export class OrderBriefViewComponent implements ResultItem, OnInit {
...
+    return (this.order.account_statement.encumbrance.quantity > 0
+            && this.order.status !== AcqOrderStatus.RECEIVED
+            && this.order.status !== AcqOrderStatus.CANCELLED);
+  }
+
+  /**
+   * Is the expenditure data should be visible or not.
+   * @returns True if the True accounting data could be visible.
+   */
+  displayExpenditureAccountingData(): boolean {
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/order/order-brief-view/order-brief-view.component.ts:100]: `get displayExpenditureAccountingData(): boolean {` ?

File: projects/admin/src/app/acquisition/components/order/place-order-form/place-order-form.component.html [hunk 12]
```diff
@@ -26,7 +26,7 @@ <h4 class="modal-title pull-left" translate>Send the order</h4>
     {{ 'You\'re about to send the order.' | translate }}<br/>
     {{ 'Below you can find the preview message that will be sent.' | translate }}
   </p>
-  <pre [innerHTML]="preview.content | nl2br"></pre>
+  <pre [innerHTML]="getPreviewContent(preview) | nl2br"></pre>
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/order/place-order-form/place-order-form.component.html:29]: Transform `preview` with a pipe ?

File: projects/admin/src/app/acquisition/components/order/place-order-form/place-order-form.component.ts [hunk 13]
```diff
@@ -79,10 +79,20 @@ export class PlaceOrderFormComponent implements OnInit {
...
   // COMPONENT FUNCTIONS ======================================================
+
+  /**
+   * Get the order preview content message.
+   * @param preview: the API response containing the preview to display.
+   * @return the content to of the preview message.
+   */
+  getPreviewContent(preview: IAcqOrderPreviewResponse): string {
+    return preview.preview.substring(this.preview.preview.indexOf('\n') + 1).trim();
+  }
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/order/place-order-form/place-order-form.component.ts:94]: Candidate to a pipe ?

File: projects/admin/src/app/acquisition/components/receipt/receipt-detail-view/receipt-detail-view.component.html [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+  You should have received a copy of the GNU Affero General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+<div *ngIf="receipt && receipt.pid && permissions" class="container">
+  <h1 class="mb-3">{{ receipt | receiptLabel }}</h1>
+  <blockquote class="receipt-note row" *ngFor="let note of receipt.notes">
+    <div class="offset-1 col-2">
+      <span class="badge badge-{{ note | noteBadgeColor }}">{{ note.type | translate }}</span>
+    </div>
+    <p class='col' [innerHTML]="note.content"></p>
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/receipt/receipt-detail-view/receipt-detail-view.component.html:24]: `class="col"`

File: projects/admin/src/app/acquisition/components/receipt/receipt-detail-view/receipt-detail-view.component.html [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+              <a [routerLink]="['/records', 'documents', 'detail', line.document.pid]" target="acq-detail">
+                {{ line.document.title_text }}
+              </a>
+              <span class="ml-auto mr-2 reception_date">{{ line.receipt_date | dateTranslate: 'shortDate' }}</span>
+            </div>
+            <div class="metadata left-offset" *ngIf="line.document.identifiers">
+              ISBN: {{ line.document.identifiers.join(', ') }}
+            </div>
+            <div class="left-offset row" *ngFor="let note of line.notes">
+              <span class="col-3 p-0"><span class="badge badge-{{ note | noteBadgeColor}}">{{ note.type | translate }}</span></span>
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/receipt/receipt-detail-view/receipt-detail-view.component.html:44]: missing space before `}`

File: projects/admin/src/app/acquisition/components/receipt/receipt-detail-view/receipt-detail-view.component.html [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    <div class="card-header total-amount row">
+      <div class="container row line no-gutters">
+        <div class="col-9" translate>Total amount</div>
+        <div class="col-2 amount">{{ receipt.total_amount | currency: receipt.currency: 'symbol' }}</div>
+      </div>
+    </div>
+  </div>
+</div>
+
+<ng-template class="" #loading>
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/receipt/receipt-detail-view/receipt-detail-view.component.html:105]: Remove `class=""`

File: projects/admin/src/app/acquisition/components/receipt/receipt-summary/receipt-summary.component.html [hunk 17]
```diff
@@ -58,7 +63,7 @@
               <div class="computation">
                 ({{ line.quantity}} x {{ line.amount | currency:receipt.currency:'symbol' }})
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/components/receipt/receipt-summary/receipt-summary.component.html:64]: missing space before `}`

File: projects/admin/src/app/acquisition/routes/receipt-lines-route.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+              }
+              return field;
+            }
+          }
+        ]
+      }
+    };
+  }
+
+
```
Comment (Garfield-fr) [inline projects/admin/src/app/acquisition/routes/receipt-lines-route.ts:64]: remove extra return

**Review by Garfield-fr — APPROVED** [review]

File: projects/shared/src/lib/component/action-button/action-button.component.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { Component, EventEmitter, Input, Output } from '@angular/core';
+
+@Component({
+  selector: 'shared-action-button',
+  template: `
+    <ng-container *ngIf="!disabled; else disabledInfo">
+      <ng-container *ngIf="url && url.length > 0; else clickButton">
+        <button type="button" class="btn btn-sm {{ class }}" [title]="title" [routerLink]="url" [queryParams]="queryParams">
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/shared/src/lib/component/action-button/action-button.component.ts:25]: I always add `class` who is a component attribute. Not need to use `ngClass` (my opinion)

**Pull request merged**